contracts/BSCValidatorSet.sol

```diff
@@ -312,107 +340,149 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
+
+  /*********************** For Temporary Maintenance **************************/
+  function canEnterMaintenance(uint256 index) public view returns (bool) {
+    if (index >= currentValidatorSet.length) {
+      return false;
     }
-    // the actually index
-    index = index - 1;
-    uint256 income = currentValidatorSet[index].incoming;
-    uint256 rest = currentValidatorSet.length - 1;
-    if (rest==0) {
-      // will not remove the validator if it is the only one validator.
-      currentValidatorSet[index].incoming = 0;
-      return;
+
+    if (
+      currentValidatorSet[index].consensusAddress == address(0)     // - 0. check if empty validator
+      || (maxNumOfMaintaining == 0 || maintainSlashScale == 0)      // - 1. check if not start
+      || numOfMaintaining >= maxNumOfMaintaining                    // - 2. check if reached upper limit
+      || !isWorkingValidator(index)                                 // - 3. check if not working(not jailed and not maintaining)
+      || validatorExtraSet[index].enterMaintenanceHeight > 0        // - 5. check if has Maintained
+      || getValidators().length <= 1                                // - 6. check num of remaining working validators
+    ) {
+      return false;
     }
-    emit validatorFelony(validator,income);
-    delete currentValidatorSetMap[validator];
-    // It is ok that the validatorSet is not in order.
-    if (index != currentValidatorSet.length-1) {
-      currentValidatorSet[index] = currentValidatorSet[currentValidatorSet.length-1];
-      currentValidatorSetMap[currentValidatorSet[index].consensusAddress] = index + 1;
+
+    return true;
+  }
+
+  function enterMaintenance() external {
+    // check maintain config
+    if (maxNumOfMaintaining == 0) {
+      maxNumOfMaintaining = INIT_MAX_NUM_OF_MAINTAINING;
     }
-    currentValidatorSet.pop();
-    uint256 averageDistribute = income/rest;
-    if (averageDistribute!=0) {
-      uint n = currentValidatorSet.length;
-      for (uint i=0;i<n;i++) {
-        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
-      }
+    if (maintainSlashScale == 0) {
+      maintainSlashScale = INIT_MAINTAIN_SLASH_SCALE;
     }
-    // averageDistribute*rest may less than income, but it is ok, the dust income will go to system reward eventually.
+
+    uint256 index = getCurrentValidatorIndex(msg.sender);
+    require(canEnterMaintenance(index), "can not enter Temporary Maintenance");
+    _enterMaintenance(msg.sender, index);
+  }
+
+  function getCurrentValidatorIndex(address _validator) public view returns (uint256) {
+    uint256 index = currentValidatorSetMap[_validator];
+    require(index > 0, "only current validators");
+
+    // the actual index
+    return index - 1;
+  }
+
+  function exitMaintenance() external {
+    uint256 index = getCurrentValidatorIndex(msg.sender);
+
+    // TODO Does it need to check jailed?
+    // require(!currentValidatorSet[index].jailed, "not jailed");
```

Comment:
Yes, I think we should not allow jailed validator to exit maintain.

Reply:
ok, comment removed